### PR TITLE
szip: conan v2 support

### PR DESCRIPTION
**Pull request by paulharris**
Please check the patch - there are a few questions around the targets.

The old recipe generated targets for `szip-static` and `szip-shared`.
It also generated an alias target for szip::szip but only for the older finders.

I want to generate the szip::szip target, and am wondering if we should continue to generate the other plain ones (and how would it be done).

Also, I get a weird error message when trying to build locally from test_package:
```
-- Using Conan toolchain: K:/env/conan3/deps/cci-fixes/szip/recipes/szip/all/test_package/build/generators/conan_toolchain.cmake
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19044.
-- Conan: Target declared 'szip::szip'
CMake Error at build/generators/szip-Target-release.cmake:34 (conan_package_library_targets):
  conan_package_library_targets Function invoked with incorrect arguments for
  function named: conan_package_library_targets
Call Stack (most recent call first):
  build/generators/szipTargets.cmake:26 (include)
  build/generators/szip-config.cmake:76 (include)
  CMakeLists.txt:5 (find_package)
```

What does it mean and how do I fix it?

### CODE REVIEW

File: recipes/szip/all/conanfile.py [hunk 1]
```diff
@@ -28,91 +31,103 @@ class SzipConan(ConanFile):
...
-        return os.path.join("lib", "cmake", "conan-official-{}-targets.cmake".format(self.name))
+# QUESTION FIXME is this required?
+#        # TODO: to remove in conan v2 once cmake_find_package* generators removed
+#        self._create_cmake_module_alias_targets(
+#            os.path.join(self.package_folder, self._module_file_rel_path),
+#            {"szip-shared" if self.options.shared else "szip-static": "szip::szip"}
+#        )
+#
+#    @staticmethod
+#    def _create_cmake_module_alias_targets(module_file, targets):
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:96]: All this is required so v1 continues to work

**Review by prince-chrismc — COMMENTED** [review]
Same feedback as the question on Slack and the issue you created :)

File: recipes/szip/all/conanfile.py [hunk 2]
```diff
@@ -28,91 +31,103 @@ class SzipConan(ConanFile):
...
+
         self.cpp_info.set_property("cmake_file_name", "szip")
-        self.cpp_info.set_property("cmake_target_name", "szip-shared" if self.options.shared else "szip-static")
-        self.cpp_info.libs = tools.collect_libs(self)
+        self.cpp_info.set_property("cmake_target_name", "szip::szip")
+
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs = ["m"]
+
+# QUESTION FIXME the old target was just "szip-shared" or "szip-static", should we keep supporting this, and how?
```
Comment (uilianries) [inline recipes/szip/all/conanfile.py:123]: Please, remove those comments. Conan V1 is still mandatory.

Comment (paulharris) [reply recipes/szip/all/conanfile.py:123]: ok

**Comment by paulharris** [comment]
I have no idea why the static build works fine, but the shared build fails -- the test can't find the header file...

File: recipes/szip/all/conanfile.py [hunk 3]
```diff
@@ -102,17 +100,23 @@ def _create_cmake_module_alias_targets(module_file, targets):
...
     @property
     def _module_file_rel_path(self):
-        return os.path.join("lib", "cmake", "conan-official-{}-targets.cmake".format(self.name))
+        return os.path.join("lib", "cmake", f"conan-official-{self.name}-targets.cmake")
 
     def package_info(self):
         self.cpp_info.set_property("cmake_file_name", "szip")
         self.cpp_info.set_property("cmake_target_name", "szip-shared" if self.options.shared else "szip-static")
-        self.cpp_info.libs = tools.collect_libs(self)
+        self.cpp_info.set_property("cmake_target_aliases", ["szip::szip"])
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:112]: This does not exist 🙈 

```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
installing this from scratch... I noticed the shared export is missing a public definition

```cmake 
set_target_properties(szip-shared PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "SZ_BUILT_AS_DYNAMIC_LIB=1"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "m"
)
```

***

full cmake for reference

```cmake
cat /home/cmcarthur/szip-2.1.1/./share/cmake/szip-targets.cmake
# Generated by CMake

if("${CMAKE_MAJOR_VERSION}.${CMAKE_MINOR_VERSION}" LESS 2.6)
   message(FATAL_ERROR "CMake >= 2.6.0 required")
endif()
cmake_policy(PUSH)
cmake_policy(VERSION 2.6...3.20)
#----------------------------------------------------------------
# Generated CMake target import file.
#----------------------------------------------------------------

# Commands may need to know the format version.
set(CMAKE_IMPORT_FILE_VERSION 1)

# Protect against multiple inclusion, which would fail when already imported targets are added once more.
set(_targetsDefined)
set(_targetsNotDefined)
set(_expectedTargets)
foreach(_expectedTarget szip-static szip-shared)
  list(APPEND _expectedTargets ${_expectedTarget})
  if(NOT TARGET ${_expectedTarget})
    list(APPEND _targetsNotDefined ${_expectedTarget})
  endif()
  if(TARGET ${_expectedTarget})
    list(APPEND _targetsDefined ${_expectedTarget})
  endif()
endforeach()
if("${_targetsDefined}" STREQUAL "${_expectedTargets}")
  unset(_targetsDefined)
  unset(_targetsNotDefined)
  unset(_expectedTargets)
  set(CMAKE_IMPORT_FILE_VERSION)
  cmake_policy(POP)
  return()
endif()
if(NOT "${_targetsDefined}" STREQUAL "")
  message(FATAL_ERROR "Some (but not all) targets in this export set were already defined.\nTargets Defined: ${_targetsDefined}\nTargets not yet defined: ${_targetsNotDefined}\n")
endif()
unset(_targetsDefined)
unset(_targetsNotDefined)
unset(_expectedTargets)


# Compute the installation prefix relative to this file.
get_filename_component(_IMPORT_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
get_filename_component(_IMPORT_PREFIX "${_IMPORT_PREFIX}" PATH)
if(_IMPORT_PREFIX STREQUAL "/")
  set(_IMPORT_PREFIX "")
endif()

# Create imported target szip-static
add_library(szip-static STATIC IMPORTED)

set_target_properties(szip-static PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "m"
)

# Create imported target szip-shared
add_library(szip-shared SHARED IMPORTED)

set_target_properties(szip-shared PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "SZ_BUILT_AS_DYNAMIC_LIB=1"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "m"
)

if(CMAKE_VERSION VERSION_LESS 2.8.12)
  message(FATAL_ERROR "This file relies on consumers using CMake 2.8.12 or greater.")
endif()

# Load information for each installed configuration.
get_filename_component(_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
file(GLOB CONFIG_FILES "${_DIR}/szip-targets-*.cmake")
foreach(f ${CONFIG_FILES})
  include(${f})
endforeach()

# Cleanup temporary variables.
set(_IMPORT_PREFIX)

# Loop over all imported files and verify that they actually exist
foreach(target ${_IMPORT_CHECK_TARGETS} )
  foreach(file ${_IMPORT_CHECK_FILES_FOR_${target}} )
    if(NOT EXISTS "${file}" )
      message(FATAL_ERROR "The imported target \"${target}\" references the file
   \"${file}\"
but this file does not exist.  Possible reasons include:
* The file was deleted, renamed, or moved to another location.
* An install or uninstall procedure did not complete successfully.
* The installation package was faulty and contained
   \"${CMAKE_CURRENT_LIST_FILE}\"
but not all the files it references.
")
    endif()
  endforeach()
  unset(_IMPORT_CHECK_FILES_FOR_${target})
endforeach()
unset(_IMPORT_CHECK_TARGETS)

# This file does not depend on other imported targets which have
# been exported from the same project but in a separate export set.

# Commands beyond this point should not need to know the version.
set(CMAKE_IMPORT_FILE_VERSION)
cmake_policy(POP)
```

File: recipes/szip/all/conanfile.py (hunk 3 above)
Comment (paulharris) [reply recipes/szip/all/conanfile.py:112]: I don't understand, what doesn't exist?   I want to be able to import szip target as szip::szip ... does this not achieve the goal?

**Comment by paulharris** [comment]
> installing this from scratch... I noticed the shared export is missing a public definition
> 
> ```cmake
> set_target_properties(szip-shared PROPERTIES
>   INTERFACE_COMPILE_DEFINITIONS "SZ_BUILT_AS_DYNAMIC_LIB=1"
>   INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"

Great eyes, thanks, I've added it

**Comment by paulharris** [comment]
Some kind of cross-compile problem?
The recipe has:
```
        if can_run(self) and self.options.enable_large_file:
            # Assume it works, otherwise raise in 'validate' function
            tc.variables["TEST_LFS_WORKS_RUN"] = True
            tc.variables["TEST_LFS_WORKS_RUN__TRYRUN_OUTPUT"] = True
```

And then the error is:

```
Make Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:
   TEST_LFS_WORKS_RUN (advanced)
   TEST_LFS_WORKS_RUN__TRYRUN_OUTPUT (advanced)
For details see /Users/jenkins/w/prod/BuildSingleReference/.conan/data/szip/2.1.1/_/_/build/a507cd2ddb8bd6879372578d0db199ba72df440f/build/Release/TryRunResults.cmake
```

**Comment by paulharris** [comment]
I think this is ready for review

File: recipes/szip/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
 find_package(szip REQUIRED CONFIG)
 
 add_executable(${PROJECT_NAME} test_package.c)
 if(TARGET szip-shared)
-    target_link_libraries(${PROJECT_NAME} szip-shared)
-else()
-    target_link_libraries(${PROJECT_NAME} szip-static)
-endif()
+   target_link_libraries(${PROJECT_NAME} szip-shared)
+else ()
```
Comment (prince-chrismc) [inline recipes/szip/all/test_package/CMakeLists.txt:10]: (insert a meme about whitespace)

**Review by prince-chrismc — APPROVED** [review]

File: recipes/szip/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (paulharris) [reply recipes/szip/all/test_package/CMakeLists.txt:10]: heh I checked the incremental commit, but you can only see that whitespace change easily with a squashed commit or a combined diff or github Changes view.

**Comment by paulharris** [comment]
I'll leave this in draft for now, I've managed to upgrade HDF5 (testing now), and szip is also used by a few other things so I was going to look at them all together.

Have I renamed the szip --> SZIP correctly ?  Any downsides?
All the finders I can find (in CCI) use find_package(SZIP) and SZIP_LIBRARIES

File: recipes/szip/all/conanfile.py [hunk 5]
```diff
@@ -28,72 +31,69 @@ class SzipConan(ConanFile):
...
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
-            tools.patch(**patch)
-        tools.replace_in_file(os.path.join(self._source_subfolder, "CMakeLists.txt"),
+        apply_conandata_patches(self)
+        replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"),
                               "set (CMAKE_POSITION_INDEPENDENT_CODE ON)", "")
-        cmake = self._configure_cmake()
+        replace_in_file(self, os.path.join(self.source_folder, "CMakeLists.txt"),
+                              "option (BUILD_SHARED_LIBS \"Build Shared Libraries\" ON)", "")
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:67]: You are adding weird things, this should be set by the toolchain and it's not force cache so I am really surprised you need this

Maybe try this secret

https://github.com/conan-io/conan-center-index/blob/843317aed0272023e4be86eb6616c645f8f9ff8e/recipes/libcorrect/all/conanfile.py#L61-L62

File: recipes/szip/all/test_v1_package/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+else()
+    target_link_libraries(${PROJECT_NAME} szip-static)
+endif()
+
+if (NOT SZIP_FOUND)
+   message (FATAL_ERROR "SZIP_FOUND should be set")
+endif ()
+
+add_executable("${PROJECT_NAME}-b" ../test_package/test_package.c)
+target_link_libraries("${PROJECT_NAME}-b" ${SZIP_LIBRARIES})
```
Comment (prince-chrismc) [inline recipes/szip/all/test_v1_package/CMakeLists.txt:21]: This is probably what's failing... the old generators do not setup ALL the possible variables this is is probably nothing

File: recipes/szip/all/conanfile.py (hunk 5 above)
Comment (prince-chrismc) [reply recipes/szip/all/conanfile.py:67]: Look at the build logs there's actually a warning saying to do this, so thats actually expected... TIL

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

File: recipes/szip/all/conanfile.py (hunk 5 above)
Comment (paulharris) [reply recipes/szip/all/conanfile.py]: Thanks, I was not enjoying adding those lines

File: recipes/szip/all/test_v1_package/CMakeLists.txt (hunk 6 above)
Comment (paulharris) [reply recipes/szip/all/test_v1_package/CMakeLists.txt:21]: Ok will remove and see

**Comment by paulharris** [comment]
For HDF5 and other existing cmake projects, they are linking to szip like:
target_link_libraries(thing ${SZIP_LIBRARIES})

I don't want to have to patch them, I'd rather support them.
It works fine in "v2" mode, just not "v1" mode.
Do I need to worry about that?
As long as the consumer recipes are updated, they'll be able to use SZIP, right?

It seems that the v1 test SOMETIMES works as expected by linking to (eg) szip-shared (on Windows it works),
but it doesn't work on Linux, I have to link to SZIP::SZIP.
The generators don't seem to export the other targets.

This is all expected, right?

File: recipes/szip/all/conanfile.py [hunk 7]
```diff
@@ -102,17 +102,25 @@ def _create_cmake_module_alias_targets(module_file, targets):
...
+        save(self, module_file, content)
 
     @property
     def _module_file_rel_path(self):
-        return os.path.join("lib", "cmake", "conan-official-{}-targets.cmake".format(self.name))
+        return os.path.join("lib", "cmake", f"conan-official-{self.name}-targets.cmake")
 
     def package_info(self):
-        self.cpp_info.set_property("cmake_file_name", "szip")
+        self.cpp_info.set_property("cmake_file_name", "SZIP")
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:112]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "szip")
```

File: recipes/szip/all/conanfile.py [hunk 8]
```diff
@@ -102,17 +102,25 @@ def _create_cmake_module_alias_targets(module_file, targets):
...
+
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.extend(["m"])
+
+        if self.options.shared:
+            self.cpp_info.defines.append("SZ_BUILT_AS_DYNAMIC_LIB=1")
 
         # TODO: to remove in conan v2 once cmake_find_package* generators removed
+        self.cpp_info.names["cmake_find_package"] = "SZIP"
+        self.cpp_info.names["cmake_find_package_multi"] = "SZIP"
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:124]:
```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
Installign this locally it gave me

```
╭─ ~/szip-2.1.1/share/cmake                                                                                                                                 ✔  07:56:07 AM
╰─ ls
szip-config-version.cmake  szip-config.cmake  szip-targets-noconfig.cmake  szip-targets.cmake
```

**Comment by prince-chrismc** [comment]
Call me crazy but their CMake never actually set it to true 🤔 


```sh
$ ╰─ cat szip-config.cmake | grep _FOUND
    if(NOT ${_NAME}_${comp}_FOUND)
        set(${_NAME}_FOUND FALSE)
      set (${SZIP_PACKAGE_NAME}_${comp}_FOUND 0)
      set (${SZIP_PACKAGE_NAME}_${comp}_FOUND 1)
```


- `cat szip-config.cmake`

```cmake 
#-----------------------------------------------------------------------------
# SZIP Config file for compiling against SZIP build directory
#-----------------------------------------------------------------------------

####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
####### Any changes to this file will be overwritten by the next CMake run ####
####### The input file was szip-config.cmake.in                            ########

get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../" ABSOLUTE)

macro(set_and_check _var _file)
  set(${_var} "${_file}")
  if(NOT EXISTS "${_file}")
    message(FATAL_ERROR "File or directory ${_file} referenced by variable ${_var} does not exist !")
  endif()
endmacro()

macro(check_required_components _NAME)
  foreach(comp ${${_NAME}_FIND_COMPONENTS})
    if(NOT ${_NAME}_${comp}_FOUND)
      if(${_NAME}_FIND_REQUIRED_${comp})
        set(${_NAME}_FOUND FALSE)
      endif()
    endif()
  endforeach()
endmacro()

####################################################################################

string(TOUPPER szip SZIP_PACKAGE_NAME)

set (${SZIP_PACKAGE_NAME}_VALID_COMPONENTS static shared)

#-----------------------------------------------------------------------------
# User Options
#-----------------------------------------------------------------------------
set (${SZIP_PACKAGE_NAME}_ENABLE_ENCODING      ON)
set (${SZIP_PACKAGE_NAME}_BUILD_SHARED_LIBS    ON)
set (${SZIP_PACKAGE_NAME}_EXPORT_LIBRARIES     szip-static;szip-shared)

#-----------------------------------------------------------------------------
# Directories
#-----------------------------------------------------------------------------
set (${SZIP_PACKAGE_NAME}_INCLUDE_DIR "${PACKAGE_PREFIX_DIR}/include")

set (${SZIP_PACKAGE_NAME}_SHARE_DIR "${PACKAGE_PREFIX_DIR}/share/cmake")
set_and_check (${SZIP_PACKAGE_NAME}_BUILD_DIR "${PACKAGE_PREFIX_DIR}")

#-----------------------------------------------------------------------------
# Version Strings
#-----------------------------------------------------------------------------
set (${SZIP_PACKAGE_NAME}_VERSION_STRING 2.1)
set (${SZIP_PACKAGE_NAME}_VERSION_MAJOR  2.1)
set (${SZIP_PACKAGE_NAME}_VERSION_MINOR  1)

#-----------------------------------------------------------------------------
# Don't include targets if this file is being picked up by another
# project which has already build SZIP as a subproject
#-----------------------------------------------------------------------------
if (NOT TARGET "szip")
  include (${PACKAGE_PREFIX_DIR}/share/cmake/szip-targets.cmake)
endif ()

# Handle default component(static) :
if (NOT ${SZIP_PACKAGE_NAME}_FIND_COMPONENTS)
    set (${SZIP_PACKAGE_NAME}_FIND_COMPONENTS static)
    set (${SZIP_PACKAGE_NAME}_FIND_REQUIRED_static true)
endif ()

# Handle requested components:
list (REMOVE_DUPLICATES ${SZIP_PACKAGE_NAME}_FIND_COMPONENTS)
foreach (comp IN LISTS ${SZIP_PACKAGE_NAME}_FIND_COMPONENTS)
    list (FIND ${SZIP_PACKAGE_NAME}_EXPORT_LIBRARIES "szip-${comp}" HAVE_COMP)
    if (${HAVE_COMP} LESS 0)
      set (${SZIP_PACKAGE_NAME}_${comp}_FOUND 0)
    else ()
      set (${SZIP_PACKAGE_NAME}_${comp}_FOUND 1)
      string(TOUPPER ${SZIP_PACKAGE_NAME}_${comp}_LIBRARY COMP_LIBRARY)
      set (${COMP_LIBRARY} ${${COMP_LIBRARY}} szip-${comp})
    endif ()
endforeach ()

check_required_components (${SZIP_PACKAGE_NAME})
```

**Comment by prince-chrismc** [comment]
```cmake
/tmp/cci-13551                                                                                                                                           ✔  09:18:01 AM
╰─ cat szip-config.cmake
########## MACROS ###########################################################################
#############################################################################################

# Requires CMake > 3.15
if(${CMAKE_VERSION} VERSION_LESS "3.15")
    message(FATAL_ERROR "The 'CMakeDeps' generator only works with CMake >= 3.15")
endif()

include(${CMAKE_CURRENT_LIST_DIR}/cmakedeps_macros.cmake)
include(${CMAKE_CURRENT_LIST_DIR}/szipTargets.cmake)
include(CMakeFindDependencyMacro)

check_build_type_defined()

foreach(_DEPENDENCY ${szip_FIND_DEPENDENCY_NAMES} )
    # Check that we have not already called a find_package with the transitive dependency
    if(NOT ${_DEPENDENCY}_FOUND)
        find_dependency(${_DEPENDENCY} REQUIRED ${${_DEPENDENCY}_FIND_MODE})
    endif()
endforeach()

set(szip_VERSION_STRING "2.1.1")
set(szip_INCLUDE_DIRS ${szip_INCLUDE_DIRS_RELEASE} )
set(szip_INCLUDE_DIR ${szip_INCLUDE_DIRS_RELEASE} )
set(szip_LIBRARIES ${szip_LIBRARIES_RELEASE} )
set(szip_DEFINITIONS ${szip_DEFINITIONS_RELEASE} )

# Only the first installed configuration is included to avoid the collision
foreach(_BUILD_MODULE ${szip_BUILD_MODULES_PATHS_RELEASE} )
    message(STATUS "Conan: Including build module from '${_BUILD_MODULE}'")
    include(${_BUILD_MODULE})
endforeach()
```

**Comment by prince-chrismc** [comment]
> For HDF5 and other existing cmake projects, they are linking to szip like:
> target_link_libraries(thing ${SZIP_LIBRARIES})

@paulharris if downstream projects are not using the correct CMake targets we should open PRs to help them improve their CMake

as for the variables being the wrong casing

https://github.com/conan-io/conan-center-index/commit/a0a72a3b7e3e0b39ee7f3be079dce34ee96d8e90

there are cases thats also true the openssl recipes has some examples for adding it for the new generators as well



(sorry for the spam)

File: recipes/szip/all/conanfile.py (hunk 8 above)
Comment (paulharris) [reply recipes/szip/all/conanfile.py:124]: I thought we still need these lines?  It is in the v2 package_template in the docs folder...

File: recipes/szip/all/conanfile.py (hunk 7 above)
Comment (paulharris) [reply recipes/szip/all/conanfile.py:112]: Why go back to lowercase?  I was aiming to follow upstream in the same way as the ZLIB library.

**Comment by paulharris** [comment]
> Call me crazy but their CMake never actually set it to true thinking

This is what drives me to the edge, I just want to get it working and move back to my own work!

**Comment by paulharris** [comment]
> @paulharris if downstream projects are not using the correct CMake targets we should open PRs to help them improve their CMake

I've done that in the past.  HDF5 in particular is active, and there are some recent github issues related to their cmake.
But as mentioned here, https://github.com/conan-io/conan-center-index/issues/13550#issuecomment-1281133018
we don't really want to set the future direction here first...

and, if we aim to support the older versions of HDF5, etc, we still need this recipe to work with the older versions of that library.
We can patch what we have in CCI, but we can't patch other projects outside of CCI.

> (sorry for the spam)

More spam = more I can get some understanding.  I still feel like I'm in the dark most of the time.


So, how do I write package_info() if I would like to support for both Conan v2 and v1:
- `find_package(szip CONFIG)` and `target_link_libraries(thingy szip::szip)`   (because it seems right)
- `find_package(szip CONFIG)` and `target_link_libraries(thingy szip-shared)` (or szip-static, depending on what was built)
- `find_package(SZIP)` and `target_link_libraries(thingy ${SZIP_LIBRARIES})`  (support legacy)
- `find_package(SZIP)` and `target_link_libraries(thingy szip-shared)` (or szip-static, depending on what was built)

The first one is the only new item in the list.

I'm pushing a commit that does all of these, AND adds some extra tests - I wonder if we can ask the CI to run them?
They had to be separate to the existing tests as I wanted to test find_package(SZIP) as opposed to find_package(szip) and I don't think they can be both in the same cmake file.
So now there is
- test_package
- test_v1_package
- test_legacy_package
- test_v1_legacy_package

Where "legacy" refers to find_package(SZIP) and SZIP_FOUND and SZIP_LIBRARIES etc ... so that we have find_package(szip) as the primary, as per @prince-chrismc  suggestion above.

What needs to change?

**Comment by paulharris** [comment]
I can't figure it how to support ${SZIP_LIBRARIES} in test_legacy_package (ie conan v2 mode).
Is it something we can support?

If not, then that means the HDF5 (and other) packages will stop working during the conan v1->v2 switch.

File: recipes/szip/all/conanfile.py [hunk 9]
```diff
@@ -102,16 +102,27 @@ def _create_cmake_module_alias_targets(module_file, targets):
                     set_property(TARGET {alias} PROPERTY INTERFACE_LINK_LIBRARIES {aliased})
                 endif()
             """.format(alias=alias, aliased=aliased))
-        tools.save(module_file, content)
+        content += """\
+                set (SZIP_FOUND TRUE)
+                set (SZIP_LIBRARIES szip::szip)
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:107]: this does not exist 😕

Comment (paulharris) [reply recipes/szip/all/conanfile.py:107]: This is my attempt to tackle this problem: https://github.com/conan-io/conan-center-index/pull/13551#issuecomment-1286358620
would appreciate some guidance ...

**Comment by prince-chrismc** [comment]
TL;DR

Dont waste time in this recipe trying to fix HDF5 - there no way to "not invent/start things here first" and make HDF5 work

Let's work on fixing the HDF5 recipe, with toolchains we have a lot more options about passing in values which should be more then enough to get (at least) less patches in HDF5 (and hopefully none)

File: recipes/szip/all/conanfile.py (hunk 9 above)
Comment (prince-chrismc) [reply recipes/szip/all/conanfile.py:107]: > So, how do I write package_info() if I would like to support for both Conan v2 and v1:
> 
> * `find_package(szip CONFIG)` and `target_link_libraries(thingy szip::szip)`   (because it seems right)
> * `find_package(szip CONFIG)` and `target_link_libraries(thingy szip-shared)` (or szip-static, depending on what was built)
> * `find_package(SZIP)` and `target_link_libraries(thingy ${SZIP_LIBRARIES})`  (support legacy)
> * `find_package(SZIP)` and `target_link_libraries(thingy szip-shared)` (or szip-static, depending on what was built)

Out of all of these only the second one is "official"

As you've pointed out, we do not want to set the direction, szip is unmainteed AFAIK so there's no way to add to it at this point

Comment (paulharris) [reply recipes/szip/all/conanfile.py:107]: Ok, so just patch HDF5 (and others) to link to whichever target is available with the `if (TARGET szip-shared)` trick in the test_package

**Comment by paulharris** [comment]
Thanks @prince-chrismc I appreciate all your work and text here.
Perhaps this can be in a policy document somewhere in the "how to make packages" and "how to upgrade for v2", because there must be quite a few of these situations.

**Comment by prince-chrismc** [comment]
> Perhaps this can be in a policy document somewhere in the "how to make packages" and "how to upgrade for v2", because there must be quite a few of these situations.

Thats my job to write those!

This has been something on my mind for a while, but I've struggled with a title. Maybe along side #13673 I will try to bring this in 💟

**Comment by paulharris** [comment]
Ok I simplified the recipe a bunch, please review.

File: recipes/szip/all/conanfile.py [hunk 10]
```diff
@@ -28,91 +31,74 @@ class SzipConan(ConanFile):
...
+        tc.generate()
 
     def package(self):
-        self.copy("COPYING", dst="licenses", src=self._source_subfolder)
-        cmake = self._configure_cmake()
+        copy(self, "COPYING", src=self.source_folder, dst=os.path.join(self.package_folder, "licenses"))
+        cmake = CMake(self)
         cmake.install()
-        # TODO: to remove in conan v2 once cmake_find_package* generators removed
-        self._create_cmake_module_alias_targets(
```
Comment (prince-chrismc) [inline recipes/szip/all/conanfile.py:90]: Quickly this should be required for the `test_v1_package` 🤔 

You should check the https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package for a reference about the 2.0 migration

Theres extras in here I am not sure about 

***

If you'd like we can do a zoom to finish this up and it will be easier to work instead of the horrible time zone gap

**Review by prince-chrismc — APPROVED** [review]

File: recipes/szip/all/conanfile.py (hunk 10 above)
Comment (paulharris) [reply recipes/szip/all/conanfile.py:90]: Chris was awesome in the zoom, and so now I'm a little bit wiser than before :)

**Review by uilianries — APPROVED** [review]
